// Copyright 2015 gRPC authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

import "src/protos/common/common.proto";

package chunkserver;

// The greeting service definition.
service Chunkserver {

  rpc AllocateChunk (AllocateChunkRequest) returns (AllocateChunkReply) {}

  rpc SendChunkData (SendChunkDataRequest) returns (SendChunkDataReply) {}

  rpc CommitChunkData (CommitChunkDataRequest) returns (CommitChunkDataReply) {}

  rpc ReadChunkData (ReadChunkDataRequest) returns (ReadChunkDataReply) {}
}

message AllocateChunkRequest {
  string chunk_handle = 1;
}

message AllocateChunkReply {
  enum AllocateChunkStatus {
    OK = 0;
    UNKNOWN = 1;
    CONFLICT = 2;
  }
  AllocateChunkStatus status = 1;
}

message SendChunkDataRequest {
  string client_id = 1;
  string chunk_handle = 2;
  int32 chunk_version =3;
  common.ByteRange range = 4;
  bytes data = 5;
}

message SendChunkDataReply {
  enum SendChunkDataStatus {
    OK = 0;
    UNKNOWN = 1;
    NOT_FOUND = 2;
    INVALID_VERSION = 3;
  }
  SendChunkDataStatus status = 1;
}

message CommitChunkDataRequest {
  string client_id = 1;
  string chunk_handle = 2;
  int32 chunk_version = 3;
}

message CommitChunkDataReply {
  enum CommitChunkDataStatus {
    OK = 0;
    UNKNOWN = 1;
    NOT_FOUND = 2;
    INVALID_VERSION = 3;
  }
  CommitChunkDataStatus status = 1;
}

message ReadChunkDataRequest {
  string chunk_handle = 1;
  common.ByteRange range = 2;
  int32 chunk_version = 3;
}

message ReadChunkDataReply {
  ReadChunkDataRequest request = 1;
  bytes data = 2;
  int32 bytes_read = 3;
  enum ReadChunkDataStatus {
    OK = 0;
    UNKNOWN = 1;
    NOT_FOUND = 2;
    INVALID_VERSION = 3;
  }
  ReadChunkDataStatus status = 4;
}
